src/mgr/DaemonServer.cc

```diff
@@ -371,64 +373,107 @@ bool DaemonServer::_allowed_command(
   return capable;
 }
 
-class ReplyOnFinish : public Context {
-  DaemonServer* mgr;
-  MCommand *m;
-  bufferlist odata;
-
-public:
-  bufferlist from_mon;
-  string outs;
-
-  ReplyOnFinish(DaemonServer* mgr, MCommand *m, bufferlist&& odata)
-    : mgr(mgr), m(m), odata(std::move(odata))
-  {}
-  void finish(int r) override {
-    odata.claim_append(from_mon);
-    mgr->_reply(m, r, outs, odata);
-  }
-};
-
 bool DaemonServer::handle_command(MCommand *m)
 {
   int r = 0;
   std::stringstream ss;
-  bufferlist odata;
   std::string prefix;
 
   assert(lock.is_locked_by_me());
 
-  cmdmap_t cmdmap;
+  /**
+   * The working data for processing an MCommand.  This lives in
+   * a class to enable passing it into other threads for processing
+   * outside of the thread/locks that called handle_command.
+   */
+  class CommandContext
+  {
+    public:
+    MCommand *m;
+    ConnectionRef con;
+    bufferlist odata;
+    cmdmap_t cmdmap;
+
+    CommandContext(MCommand *m_)
+      : m(m_)
+    {
+      con = m->get_connection();
+    }
+
+    ~CommandContext()
+    {
+      m->put();
+    }
+
+    void reply(int r, const std::stringstream &ss)
+    {
+      reply(r, ss.str());
+    }
+
+    void reply(int r, const std::string &rs)
+    {
+      // Let the connection drop as soon as we've sent our response
+      if (m->get_connection()) {
+        m->get_connection()->mark_disposable();
+      }
+
+      dout(1) << "do_command r=" << r << " " << rs << dendl;
+      if (con) {
```

Comment:
why are we using the `con` stored when the `CommandContext` instance is created instead of the one we are checking at line 416? if `! m->get_connection()`, i don't think we should reply the client anymore.